// vue3-lazy-hydration@1.2.1/dist/esm/vue3-lazy-hydration.mjs
import { default as default2 } from "./components/LazyHydrationWrapper.mjs";
import { default as default3 } from "./composables/useLazyHydration.mjs";
import { default as default4 } from "./composables/useHydrateWhenIdle.mjs";
import { default as default5 } from "./composables/useHydrateOnInteraction.mjs";
import { default as default6 } from "./composables/useHydrateWhenTriggered.mjs";
import { default as default7 } from "./composables/useHydrateWhenVisible.mjs";
import { default as default8 } from "./wrappers/hydrate-never.mjs";
import { default as default9 } from "./wrappers/hydrate-when-idle.mjs";
import { default as default10 } from "./wrappers/hydrate-when-visible.mjs";
import { default as default11 } from "./wrappers/hydrate-on-interaction.mjs";
import { default as default12 } from "./wrappers/hydrate-when-triggered.mjs";
export {
  default2 as LazyHydrationWrapper,
  default8 as hydrateNever,
  default11 as hydrateOnInteraction,
  default9 as hydrateWhenIdle,
  default12 as hydrateWhenTriggered,
  default10 as hydrateWhenVisible,
  default5 as useHydrateOnInteraction,
  default4 as useHydrateWhenIdle,
  default6 as useHydrateWhenTriggered,
  default7 as useHydrateWhenVisible,
  default3 as useLazyHydration
};
