// eslint-compat-utils@0.2.1/dist/linter.cjs
'use strict';

const eslint = require('eslint');
const semver = require('semver');
const convertConfig = require('./shared/eslint-compat-utils.95f935d7.cjs');
require('module');

function _interopNamespaceCompat(e) {
  if (e && typeof e === 'object' && 'default' in e) return e;
  const n = Object.create(null);
  if (e) {
    for (const k in e) {
      n[k] = e[k];
    }
  }
  n.default = e;
  return n;
}

const eslint__namespace = /*#__PURE__*/_interopNamespaceCompat(eslint);
const semver__namespace = /*#__PURE__*/_interopNamespaceCompat(semver);

let cacheLinter;
function getLinter() {
  return cacheLinter != null ? cacheLinter : cacheLinter = getLinterInternal();
  function getLinterInternal() {
    if (semver__namespace.gte(eslint__namespace.Linter.version, "9.0.0-0")) {
      return eslint__namespace.Linter;
    }
    return getLinterClassForV8();
  }
}
function getLinterClassForV8() {
  return class LinterForV8 extends eslint__namespace.Linter {
    static get version() {
      return eslint__namespace.Linter.version;
    }
    verify(code, config, option) {
      const newConfig = convertConfig.convertConfigToRc(config, this);
      return super.verify(code, newConfig, option);
    }
  };
}

exports.getLinter = getLinter;
